File: tests/lexer_test.nim

import unittest
import strutils, sequtils
import compiler.fcore, compiler.internalpass

initFlori()

proc lex(src: string): seq[Token] =
  lex("testmodule.flori", src)

suite "F fxpression lexer test":
  test "block test":
    let tokens = lex("""
add5 =>
  x + 5
""")
    check tokens[0].kind == tokenIdent
    check tokens[1].kind == tokenInfix
    check tokens[2].kind == tokenNewline
    check tokens[3].kind == tokenLBlock
    check tokens[^1].kind == tokenRBlock
  test "fib test":
    let tokens = lex("""
fib =>
  if n<2
    n
  else
    fib(n-1) + fib(n-2)
""")
    check tokens[0].kind == tokenIdent
    check $tokens[0] == "fib"
    check $tokens[1] == "=>"
    check tokens[2].kind == tokenNewline
    check tokens[3].kind == tokenLBlock
    check tokens[4].kind == tokenIdent
    check tokens[5].kind == tokenIdent
    check tokens[6].kind == tokenInfix
    check tokens[7].kind == tokenIntLit
    check tokens[8].kind == tokenNewline
    check tokens[9].kind == tokenLBlock
    check tokens[10].kind == tokenIdent
    check tokens[11].kind == tokenRBlock
    check tokens[12].kind == tokenNewline

File: test/lexer_test.c
#include <assert.h>
#include <string.h>
#include "flori.h"

int main() {
  lexer* lx = new_lexer(stdin, "stdin.flori");
  tokenstream* ts = lex(lx);
  for (;;) {
    token* t = next_token(ts);
    if (t == NULL) break;
    if (t->kind == TOKEN_INTLIT) {
      printf("%s:%d ", tokenkind_tostring(t->kind), t->intval);
    } else if (t->kind == TOKEN_IDENT) {
      printf("%s:%s ", tokenkind_tostring(t->kind), t->ident);
    } else if (t->kind == TOKEN_OP) {
      printf("%s:%s ", tokenkind_tostring(t->kind), t->ident);
    } else {
      printf("%s ", tokenkind_tostring(t->kind));
    }
  }
}
